crates/matrix-sdk/src/widget_api/driver/mod.rs

```diff
@@ -0,0 +1,231 @@
+use tokio::sync::mpsc;
+
+pub mod widget;
+
+#[derive(Debug)]
+pub struct Driver<W: Widget> {
+    pub matrix_room: Joined,
+    pub widget: W,
+    add_event_handler_handle: Option<EventHandlerHandle>,
+}
+
+#[async_trait(?Send)]
+impl<W: Widget> handler::Driver for Driver<W> {
+    async fn send(&self, _message: handler::Outgoing) -> Result<()> {
+        // let message_str = serde_json::to_string(&message)?;
+        let _ = self.widget.send_widget_message("TODO get message string from outgoing");
+        Result::Ok(())
+    }
+    async fn initialise(&mut self, options: Options) -> Result<Capabilities> {
+        let options = self.widget.capability_permissions(options).await?;
+
+        let mut capabilities = Capabilities::default();
+
+        capabilities.event_listener = self.build_event_listener(&options.read_filter);
+        capabilities.event_sender = self.build_event_sender(&options.send_filter);
+        capabilities.event_reader = self.build_event_reader(&options.read_filter);
+        Result::Ok(capabilities)
+    }
+    async fn get_openid(&self, req: openid::Request) -> OpenIDState {
+        // TODO: make the client ask the user first.
+        // We wont care about this for Element call -> Element X
+        // if !self.has_open_id_user_permission() {
+        //     let (rx,tx) = tokio::oneshot::channel();
+        //     return OpenIDState::Pending(tx);
+        //     let permissions = widget.openid_permissions().await?;
+        //     if (permissions.should_cache){
+        //        self.auto_allow_openid_token(permission.is_allowed);
+        //     }
+        //     self.get_openid(req, Some(tx)); // get open id can be called with or without tx and either reutrns as return or sends return val over tx
+        // }
+
+        let user_id = self.matrix_room.client.user_id();
+        if user_id == None {
+            return OpenIDState::Resolved(Err(WidgetError(
+                "Failed to get an open id token from the homeserver. Because the userId is not available".to_owned()
+            )));
+        }
+        let user_id = user_id.unwrap();
+
+        let request =
+            ruma::api::client::account::request_openid_token::v3::Request::new(user_id.to_owned());
+        let res = self.matrix_room.client.send(request, None).await;
+
+        let state = match res {
+            Err(err) => Err(WidgetError(
+                format!(
+                    "Failed to get an open id token from the homeserver. Because of Http Error: {}",
+                    err.to_string()
+                )
+                .to_owned(),
+            )),
+            Ok(res) => Ok(openid::Response {
+                id: req.id,
+                token: res.access_token,
+                expires_in_seconds: res.expires_in.as_secs() as usize,
+                server: res.matrix_server_name.to_string(),
+                kind: res.token_type.to_string(),
+            }),
+        };
+        OpenIDState::Resolved(state)
```

Comment:
A more elegant way of writing it is:

```suggestion
let res = self.room.client.send(request, None).await.map_err(|e| WidgetError(e.to_string()))?;
Ok(OpenIDState::Resolved(openid::Response {
    id: req.id,
    token: res.access_token,
    expires_in_seconds: res.expires_in.as_secs() as usize,
    server: res.matrix_server_name.to_string(),
    kind: res.token_type.to_string(),
}));
```

Reply:
Awsome!

Reply:
This changes the structure right?
From 
```
State{
  Resolve {Result<Response>}
  Pending {}
}
```
To
```
Result<State{
  Resolve {Response}
  Pending {}
}>
```

Reply:
comment from daniel:
> - Define #[from] macro on error::Error trait that corresponds to the error that the SDK returns, this would allow you to use ? conversion.
> - Move this piece of code in a tiny function.
> - This would then allow you to make this conversion as a one liner.